File: src/org/mtransit/parser/gtfs/data/GTripStop.java
package org.mtransit.parser.gtfs.data;

import org.mtransit.parser.Constants;

// https://developers.google.com/transit/gtfs/reference#stop_times_trip_id_field
// https://developers.google.com/transit/gtfs/reference#stop_times_stop_id_field
public class GTripStop {

	private String uid;

	public static final String TRIP_ID = "trip_id";
	private String trip_id;
	public static final String STOP_ID = "stop_id";
	private String stop_id;
	public static final String STOP_SEQUENCE = "stop_sequence";
	private int stop_sequence;

	public GTripStop(String uid, String trip_id, String stop_id, int stop_sequence) {
		this.uid = uid;
		this.trip_id = trip_id;
		this.stop_id = stop_id;
		this.stop_sequence = stop_sequence;
	}

	public String getUID() {
		return this.uid;
	}

	public String getTripId() {
		return this.trip_id;
	}

	public String getStopId() {
		return this.stop_id;
	}

	public int getStopSequence() {
		return this.stop_sequence;
	}

	@Override
	public String toString() {
		return new StringBuilder() //
				.append(Constants.STRING_DELIMITER).append(this.uid).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.trip_id).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.stop_id).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.append(Constants.STRING_DELIMITER).append(this.stop_sequence).append(Constants.STRING_DELIMITER).append(Constants.COLUMN_SEPARATOR) //
				.toString();
	}

	public static String getUID(String trip_uid, String stop_id, int stop_sequence) {
		return stop_id + Constants.UUID_SEPARATOR + stop_sequence + Constants.UUID_SEPARATOR + trip_uid;
	}
}

File: src/main/java/org/mtransit/parser/gtfs/data/GTripStop.kt
package org.mtransit.parser.gtfs.data

import androidx.annotation.Discouraged

// https://developers.google.com/transit/gtfs/reference#stop_timestxt
// https://gtfs.org/reference/static#stop_timestxt
// -_trip_id field
// - stop_id field
data class GTripStop(
    val routeIdInt: Int,
    val tripIdInt: Int,
    val stopIdInt: Int,
    val stopSequence: Int
) {

    @Suppress("unused")
    constructor(
        routeAndTripUID: String,
        stopId: Int,
        stopSequence: Int
    ) : this(
        GTrip.split(routeAndTripUID),
        stopId,
        stopSequence
    )

    @Suppress("unused")
    constructor(
        routeAndTripUID: String,
        tripIdInt: Int,
        stopId: Int,
        stopSequence: Int
    ) : this(
        GTrip.extractRouteIdInt(routeAndTripUID),
        tripIdInt,
        stopId,
        stopSequence
    )

    constructor(
        routeAndTripUID: Pair<Int, Int>,
        stopId: Int,
        stopSequence: Int
    ) : this(
        routeAndTripUID.first,
        routeAndTripUID.second,
        stopId,
        stopSequence
    )

    val uID by lazy { getNewUID(routeIdInt, tripIdInt, stopIdInt, stopSequence) }

    @Discouraged(message = "Not memory efficient")
    @Suppress("unused")
    val tripId = _tripId

    @Suppress("unused")
    private val _tripId: String
        get() {
            return GIDs.getString(tripIdInt)
        }

    @Discouraged(message = "Not memory efficient")
    @Suppress("unused")
    val stopId = _stopId

    private val _stopId: String
        get() {
            return GIDs.getString(stopIdInt)
        }

    @Suppress("unused")
    fun toStringPlus(): String {
        return toString() +
                "+(stopId:$_stopId)" +
                "+(tripId:$_tripId)"
    }

    companion object {
        const val ROUTE_ID = GRoute.ROUTE_ID
        const val TRIP_ID = GTrip.TRIP_ID
        const val STOP_ID = GStop.STOP_ID
        const val STOP_SEQUENCE = "stop_sequence"

        private const val UID_SEPARATOR = "+" // int IDs can be negative

        @JvmStatic
        fun getNewUID(
            tripUID: String,
            stopIdInt: Int,
            stopSequence: Int
        ) = "${tripUID}$UID_SEPARATOR${stopIdInt}$UID_SEPARATOR${stopSequence}"

        @JvmStatic
        fun getNewUID(
            routeIdInt: Int,
            tripIdInt: Int,
            stopIdInt: Int,
            stopSequence: Int
        ) = "${routeIdInt}$UID_SEPARATOR${tripIdInt}$UID_SEPARATOR${stopIdInt}$UID_SEPARATOR${stopSequence}"

        @Suppress("unused")
        @JvmStatic
        fun toStringPlus(gTripStops: Iterable<GTripStop>): String {
            return gTripStops.joinToString { it.toStringPlus() }
        }
    }
}